Resources/doc/configuration.md
Configuration
-------------

### Configuration reference

Here you can see the default configuration of the API key bundle.

``` yaml
# Default configuration for "Ma27ApiKeyAuthenticationBundle"
ma27_api_key_authentication:
    user:
        api_key_length:       200
        object_manager:       ~ # Required
        model_name:           AppBundle\Entity\User
        password:
            strategy:             php55
        metadata_cache:       false
    api_key_purge:
        enabled:              false
        last_action_listener:
            enabled:              true
        outdated_rule:        '-5 days'
    services:
        auth_handler:         null
        key_factory:          null
    key_header:           X-API-KEY
    response:
        api_key_property:     apiKey
        error_property:       message
```

### Explanations

#### `user` section

- The `api_key_length` value is the length represents the keylength of the API key.
  This value should be an even number since it the generator uses `bin2hex` to normalize the API key before saving.

- The `object_manager` is the service name of the `ObjectManager` implemented by any Doctrine implementation.

- The `model_name` is the FQCN of the user entity to be managed.

- The `password.strategy` is the alias of the hashing strategy to be used during the authentication.

- The `metadata_cache` is a flag whether or not to cache the evaluated Metadata.

#### `api_key_purge` section

- The `enabled` is a flag whether or not to enable the API key cleanup tool.

- The `last_action_listener.enabled` is a flag whether or not to enable a listener which updates the `lastAction`
  value in a user entity during the authentication process.

- The `outdated_rule` is a PHP/datetime rule which determines how old the latest action of a user must be until the Session cleanup
  removes the API key from a user.

#### `services` section

- The `auth_handler` is optional and may contain a service ID which will be used instead of the actual `ApiKeyAuthenticationHandler`.

- The `key_factory` is optional and may contain a service ID which will be used instead of the actual `KeyFactory`.

#### `key_header` section

The `key_header` is a simple, scalar value which contains the name of the API key header. Default is `X-API-KEY`.

#### `response` section

- The `api_key_property` is the name of the property in the JSON response which contains the requested API key. Default is `apiKey`.

- The `error_property` is the name of the property in the JSON response which contains the error message from the API key authentication. Default is `message`.

### [Next: Basic usage](https://github.com/Ma27/Ma27ApiKeyAuthenticationBundle/blob/master/Resources/doc/basic-usage.md)


CHANGELOG.md
# Changelog for 1.x version

## 2.0.0

- [minor] used the PHP 7.x `random_bytes()` API rather than the `openssl_` function for API key and salt generation.

- [breaking] `doctrine/common` must be at least `v2.6`.

- [breaking] `ApiKeyAuthenticationHandler#getClassMetadata` has been made protected.

- [breaking] made `hautelook/phpass` a dev dependency to avoid unnecessary installs.

- [breaking] dropped support for `hhvm` and all PHP versions lower than PHP 7.1

- [feature] Made `KeyFactory#doGenerate` protected in order to override only the generation part.

- [minor] The length computation `actual_length / 2` for the random key will be done in the `KeyFactory` instead of the extension.

- [breaking] Removed the configuration option `password.phpass_iteration_length`. Instead the parameter `ma27_api_key_authentication.password_hasher.phpass.iteration_length` should be modified manually.

- [breaking] Removed the configuration option `services.password_hasher`. Create a custom hasher via DI tags instead.

- [breaking] Removed the `sha512` strategy and the `Ma27\ApiKeyAuthenticationBundle\Model\Password\Sha512PasswordHasher` as it didn't provide an appropriate salting strategy which made the whole implementation fragile.

- [minor] Added the `ma27_api_key_authentication.password_hasher.php55.cost`

- [breaking] removed `Ma27\ApiKeyAuthenticationBundle\Security\ApiKeyAuthenticator::API_KEY_HEADER`. The API key header is configurable, so no constant is needed anymore.

- [breaking] removed `Ma27\ApiKeyAuthenticationBundle\Event\AbstractUserEvent::isUserAvailable`. The method is only available in the subclass `Ma27\ApiKeyAuthenticationBundle\Event\OnInvalidCredentialsEvent` and in other events not needed.

- [breaking] removed the configuration option `api_key_purge.log_state` and `api_key_purge.logger`. In order to tackle logger support, please use custom event hooks.

- [feature] `user.password` configuration can be omitted if the `php55` strategy is used.

- [feature] the datetime expression for the session cleanup can be configured with the configuration option `api_key_purge.outdated_rule`.

- [breaking] renamed `Ma27\ApiKeyAuthenticationBundle\Event\AssembleResponseEvent` to `Ma27\ApiKeyAuthenticationBundle\Event\OnAssembleResponseEvent` to keep the naming convention for event objects.

- [breaking] removed the `ma27_api_key_authentication.cleanup.complete` event as it will be called right after the success event and has no real benefit.

- [breaking] moved model classes to a new namespace:
  - `Ma27\ApiKeyAuthenticationBundle\Model\Key\KeyFactory` => `Ma27\ApiKeyAuthenticationBundle\Service\Key\KeyFactory`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Key\KeyFactoryInterface` => `Ma27\ApiKeyAuthenticationBundle\Service\Key\KeyFactoryInterface`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Login\ApiToken\ApiKeyAuthenticationHandler` => `Ma27\ApiKeyAuthenticationBundle\Service\Auth\ApiKeyAuthenticationBundle`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Login\AuthenticationHandlerInterface` => `Ma27\ApiKeyAuthenticationBundle\Service\Auth\AuthenticationHandlerInterface`
  - `Ma27\ApiKeyAuthenticationBundle\Model\User\ClassMetadata` => `Ma27\ApiKeyAuthenticationBundle\Service\Mapping\ClassMetadata`
  - `Ma27\ApiKeyAuthenticationBundle\Model\User\ModelConfigurationDriverInterface` => `Ma27\ApiKeyAuthenticationBundle\Service\Mapping\ModelConfigurationDriverInterface`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Password\PasswordHasherInterface` => `Ma27\ApiKeyAuthenticationBundle\Service\Password\PasswordHasherInterface`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Password\CryptPasswordHasher` => `Ma27\ApiKeyAuthenticationBundle\Service\Password\CryptPasswordHasher`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Password\PHPassHasher` => `Ma27\ApiKeyAuthenticationBundle\Service\Password\PHPassHasher`
  - `Ma27\ApiKeyAuthenticationBundle\Model\Password\PhpPasswordHasher` => `Ma27\ApiKeyAuthenticationBundle\Service\Password\PHPPasswordHasher`

- [feature] added `user.metadata_cache` option which can be used to enable a cache for the metadata.
  (__NOTE:__ certain implementation details of the metadata API were changed, but not listed as the whole metadata API is private and tagged with the `@internal` tag)

## 1.2.1

- [bug] `33e4bfe` invalid import leads to issues when dumping the DI graph

## 1.2.0

- [feature] made hashing services configurable: (#33)
  - created tag `ma27_api_key_authentication.password_hasher` which allows to create custom services and use them via the configuration
  - no BC breaks (all currently existing hashers can be used by the same config)

- [bug] ensure that no api key is generated before a new one will be created (#48)

- [feature] added `response.api_key_property` and `response.error_property` to the configuration to keep the response configurable (#45)

- [minor] deprecated `api_key_purge.logger_service`, logger support will be removed in 2.0 (#50)

- [feature] added a listener which updates the `last_action` during the firewall login and after the api key request (#55)

- [feature] added an event `ma27_api_key_authentication.credential_exception_thrown` that will be always triggered when a `CredentialException` is thrown (#56)

- [minor] deprecated `services.password_hasher`, instead a custom password hasher should be created.

## 1.1.0

- [feature] made api_key header configurable: (#32)
  - added option `key_header` to make the header name mutable.
  - deprecated the const `Ma27\ApiKeyAuthenticationBundle\Security\ApiKeyAuthenticator::API_KEY_HEADER`.

- [minor] deprecated `Ma27\ApiKeyAuthenticationBundle\Event\AbstractUserEvent#isUserAvailable`: (#36)
  - the method has been moved to `Ma27\ApiKeyAuthenticationBundle\Event\OnInvalidCredentialsEvent` since it is only needed in this subclass of the base event class.
  - the old method can be used, but triggers is declared as `deprecated`.


DependencyInjection/Configuration.php
<?php

namespace Ma27\ApiKeyAuthenticationBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from your app/config files.
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('ma27_api_key_authentication');

        $rootNode
            ->children()
                ->arrayNode('user')
                    ->children()
                        ->integerNode('api_key_length')
                            ->min(50)
                            ->defaultValue(200)
                        ->end()
                        ->scalarNode('object_manager')->isRequired()->end()
                        ->scalarNode('model_name')->defaultValue('AppBundle\\Entity\\User')->end()
                        ->arrayNode('password')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('strategy')
                                    ->defaultValue('php55')
                                ->end()
                            ->end()
                        ->end()
                        ->scalarNode('metadata_cache')->defaultFalse()->end()
                    ->end()
                ->end()
                ->arrayNode('api_key_purge')
                    ->canBeEnabled()
                    ->children()
                        ->arrayNode('last_action_listener')
                            ->canBeDisabled()
                        ->end()
                        ->scalarNode('outdated_rule')->defaultValue('-5 days')->end()
                    ->end()
                ->end()
                ->arrayNode('services')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('auth_handler')->defaultNull()->end()
                        ->scalarNode('key_factory')->defaultNull()->end()
                    ->end()
                ->end()
                ->scalarNode('key_header')->defaultValue('X-API-KEY')->end()
                ->arrayNode('response')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('api_key_property')->defaultValue('apiKey')->end()
                        ->scalarNode('error_property')->defaultValue('message')->end()
                    ->end()
                ->end()
            ->end();

        return $treeBuilder;
    }
}
